crates/sp-lightclient/src/lib.rs

```diff
@@ -496,6 +538,116 @@ impl<Header: HeaderT, Store: Storage<Header>> HeaderImporter<Header, Store> {
+            records_roots_count += digest_items.records_roots.len() as u64;
+
+            header = self
+                .store
+                .header(*header.header.parent_hash())
+                .ok_or_else(|| ImportError::MissingParent(header.header.hash()))?;
+        }
+
+        Ok(records_roots_count * u64::from(MERKLE_NUM_LEAVES))
+    }
+
+    /// Finds a records root mapped against a segment index in the chain with chain_tip as the tip of the chain.
+    /// We try to find the records root as follows,
+    ///  - Find records root from the store and return if found.
+    ///  - Find records root from the genesis record roots and return if found.
+    ///  - Find the records root present in the non finalized headers.
+    fn find_records_root_for_segment_index(
+        &self,
+        segment_index: SegmentIndex,
+        chain_tip: HashOf<Header>,
+    ) -> Result<RecordsRoot, ImportError<Header>> {
+        // check if the records root is already in the store
+        if let Some(records_root) = self.store.records_root(segment_index) {
+            return Ok(records_root);
+        };
+
+        // special case: check the genesis records roots if the Block #1 is not finalized yet
+        if let Some(records_root) = self
+            .store
+            .chain_constants()
+            .genesis_records_roots
+            .get(&segment_index)
+        {
+            return Ok(*records_root);
+        }
+
+        // find the records root from the headers which are not finalized yet.
+        let finalized_header = self.store.finalized_header();
+        let mut header = self
+            .store
+            .header(chain_tip)
+            .ok_or(ImportError::MissingHeader(chain_tip))?;
+
+        while header.header.hash() != finalized_header.header.hash() {
+            let digest_items = extract_subspace_digest_items::<
+                _,
+                FarmerPublicKey,
+                FarmerPublicKey,
+                FarmerSignature,
+            >(&header.header)?;
+
+            if let Some(records_root) = digest_items.records_roots.get(&segment_index) {
+                return Ok(*records_root);
+            }
+
+            header = self
+                .store
+                .header(*header.header.parent_hash())
+                .ok_or_else(|| ImportError::MissingParent(header.header.hash()))?;
+        }
+
+        Err(ImportError::MissingRecordsRoot(segment_index))
+    }
+
+    /// Stores finalized header and records roots present in the header.
+    fn store_finalized_header_and_records_roots(
+        &mut self,
+        header: &Header,
+    ) -> Result<(), ImportError<Header>> {
+        // mark header as finalized
+        self.store.finalize_header(header.hash());
+
+        // store the records roots present in the header digests
+        let digests_items =
+            extract_subspace_digest_items::<_, FarmerPublicKey, FarmerPublicKey, FarmerSignature>(
+                header,
+            )?;
```

Comment:
Probably nicer to move this extraction before the line`self.store.finalize_header(header.hash());` so that the operation of `finalize_header` and `store_records_roots` can be finished without an interruption in theory.